Allow SplitContainer to have more than two children

kitbdev opened the pull request:
- Closes https://github.com/godotengine/godot-proposals/issues/9666
- Related previous PR #83359
- Needed for #90439

SplitContainers can now work with more than two children. A dragger will be added between each child and they can all be moved independently. Moving a dragger into another one will push that one.
`split_offset` is deprecated in favor of a PackedInt32Array `split_offsets` so all split offsets can be adjusted. Same with `get_drag_area_controls`.
This should not break compatibility.

Split offsets are offsets from their default positions (formerly `wished_size`). The default positions are calculated similar to how BoxContainers sort their children. However, if the dragger is before or after all expand flags, then it will be at the start or end of the SplitContainer to keep the same behavior as before.

The number of split offsets automatically adjusts to match the number of valid children in the SplitContainer. Valid children are now maintained in SplitContainer and are visible non-top-level not internal Controls.

Adding and removing valid children (including changing visibility) will update the other split offsets to account for it. Children will attempt to be added at their current size.

Also adds tests for multiple children in SplitContainer.

There is a separate PR to have the editor take advantage of this: https://github.com/godotengine/godot/pull/90439

https://github.com/user-attachments/assets/908b74df-6f79-4d08-aef1-2e148c04fd8d

kitbdev commented:
> For that, you could use an `_undo_redo_inspector_callback`

I can't use it in split_container.cpp, right? Undoing adding a child should undo adding the split_offset, but it happens in the sort children notification, so I don't know how to add it to undo redo.

> Hmm, maybe I misunderstand how things are implemented, but I believe we should avoid that. I don't remember exactly why, but this was how it used to be implemented before and it created problems when resizing the container, or when children components change their minimum size.

If I didn't account for the minimum size, then the children with expand flags would not be the same proportions with their stretch ratio.
For example (B and D set to expand and a stretch ratio of 1):

![image](https://github.com/godotengine/godot/assets/10054226/e94b562b-e832-4784-bbea-4d572662b4d0)

If the minimum size of A increases, then B and D wouldn't be the same size any more. So C should be moved to keep them the at same ratio.

I don't think this is a problem since it only affects draggers that are in between two expanding children, so its default position will only be moved to keep them at the same ratio.

groud commented:
> I can't use it in split_container.cpp, right? Undoing adding a child should undo adding the split_offset, but it happens in the sort children notification, so I don't know how to add it to undo redo.

Ah right. This has been a problem forever... A ton of properties (size flags, stretch ratio, etc...) are in the Control node, in case they end up used as a child of a container.  But often they end up unused. 
Here, the "right" solution would be to add properties to Control, but that still sucks... I guess your current implementation is the most acceptable for now then :/

> Removed the SplitContainerDragger because it is easier to just manage an vector of rects than creating a new one each time a child is added and removing when a child is deleted.

This does not sound right, it was added by https://github.com/godotengine/godot/pull/65355/ to allow the dragger to have a bigger area than the underlying SplitContainer node.

TBH,  I feel this is a lot of changes. I can't really wrap my head around all changes done there. So I believe this should be split into a core PR (SplitContainer) and an editor PR (editor changes). SplitContainer has been the source of tons of small issues in the past, so we need to test the new behavior thoroughly.

Calinou commented:
Would this be useful to implement https://github.com/godotengine/godot-proposals/issues/9676?

kitbdev commented:
> Would this be useful to implement [godotengine/godot-proposals#9676](https://github.com/godotengine/godot-proposals/issues/9676)?

Yes, this (along with https://github.com/godotengine/godot/pull/90439) will make it easier to implement.
`move_child` could be used instead of removing and adding children to specific nested SplitContainers.

groud commented on scene/gui/split_container.cpp:
```diff
@@ -39,7 +39,7 @@ void SplitContainerDragger::gui_input(const Ref<InputEvent> &p_event) {
 	SplitContainer *sc = Object::cast_to<SplitContainer>(get_parent());
 
-	if (sc->collapsed || !sc->get_containable_child(0) || !sc->get_containable_child(1) || sc->dragger_visibility != SplitContainer::DRAGGER_VISIBLE) {
+	if (sc->collapsed || !sc->get_containable_child(1) || sc->dragger_visibility != SplitContainer::DRAGGER_VISIBLE) {
```
> sc->get_containable_child(1)

This seems like it is still hardcoded for two children.

kitbdev replied:
This actually checks for at least two children. When there is 0 or 1 children, `get_containable_child(1)` will be false and when there is 2 or more it will be true.
I changed this in a few places, but `get_containable_child(1)` implies `get_containable_child(0)`.

groud replied:
Ah ok that makes sense. In such case, I think it would be good to write a comment. Or maybe be explicit about it by defining a variable instead:
```
bool has_at_least_two_containable_child = sc->get_containable_child(1);
if (sc->collapsed || has_at_least_two_containable_child || sc->dragger_visibility != SplitContainer::DRAGGER_VISIBLE) {
 ...
```
Something like that I'd say.

kitbdev commented:
I found a crash when duplicating SplitContainers with multiple children, I think it's related to how I dynamically add SplitContainerDragger as internal children. I add and remove them in the sort children notification.
Is there anything I can do so draggers aren't duplicated?

kitbdev commented:
Rebased and fixed conflicts.

When duplicating a SplitContainer with 3+ children I'm getting:
` scene\main\node.cpp:1682 - Index p_index = 4 is out of bounds ((int)data.children_cache.size() = 4).`
`Child node disappeared while duplicating.`
so I may need to find a different way to handle the draggers that get added and removed dynamically.

KoBeWi commented:
Do draggers need to be nodes? 🤔

kitbdev commented:
Yes, since the grab area needs to be able to be bigger than the Split Container.
See https://github.com/godotengine/godot/pull/90411#issuecomment-2045484023
I changed them originally, but I had to change it back to Nodes.

Maybe there could be just one dragger handler? It would complicate the dragging logic but it might work.

Mickeon commented:
What's the consensus on this? It's marked for 4.3 and even though it's a nice feature, we are in feature freeze right now.

kitbdev commented:
Unfortunately there are some issues with this and I'm not sure how to fix it. (https://github.com/godotengine/godot/pull/90411#issuecomment-2133992173)
It has to do with the dynamically added draggers getting added/removed while the SplitContainer is being duplicated.
Also there can't just be one since it would take input from the controls in between. I also don't want to change the order that child nodes are duplicated in, and they need to be at the end so they are on top.

This has the potential to break things a lot with the Editor so I don't expect https://github.com/godotengine/godot/pull/90439 to be merged in 4.3 even if this was fixed now.
I'm also spending more time on bugfixes, so this probably won't be ready for 4.3 in time.

Mickeon commented on doc/classes/SplitContainer.xml:
```diff
@@ -36,39 +41,39 @@
 		</theme_item>
 		<theme_item name="separation" data_type="constant" type="int" default="12">
-			The space between sides of the container.
+			The size of the space in between the child [Control]s.
```
```suggestion
			The size of the space in between the child [Control] nodes.
```

Mickeon commented on doc/classes/SplitContainer.xml:
```diff
@@ -13,19 +13,24 @@
+			The first element of [member split_offsets].
+		</member>
+		<member name="split_offsets" type="PackedInt32Array" setter="set_split_offsets" getter="get_split_offsets" default="PackedInt32Array(0)">
+			Offsets for each dragger in pixels. Each one is the offset of the split between the [Control]s before and after the dragger, with [code]0[/code] being the default position. The default position is based on the [Control]s expand flags and minimum sizes. See [member Control.size_flags_horizontal], [member Control.size_flags_vertical], and [member Control.size_flags_stretch_ratio].
```
```suggestion
			Offsets for each dragger in pixels. Each one is the offset of the split between the [Control] nodes before and after the dragger, with [code]0[/code] being the default position. The default position is based on the [Control] nodes expand flags and minimum sizes. See [member Control.size_flags_horizontal], [member Control.size_flags_vertical], and [member Control.size_flags_stretch_ratio].
```

Mickeon commented on doc/classes/SplitContainer.xml:
```diff
@@ -13,19 +13,24 @@
 	<members>
 		<member name="collapsed" type="bool" setter="set_collapsed" getter="is_collapsed" default="false">
-			If [code]true[/code], the area of the first [Control] will be collapsed and the dragger will be disabled.
+			If [code]true[/code], the [Control]s will be set to their default positions as if all [member split_offsets] were set [code]0[/code] and the draggers will be disabled.
```
```suggestion
			If [code]true[/code], the [Control] nodes will be set to their default positions (as if all [member split_offsets] were set [code]0[/code]) and the draggers are disabled.
```

AThousandShips replied:
Also `set to [code]0[/code]`

kitbdev commented:
Rebased.
Applied suggestions.

- Found a fix to the duplication error I was having, it looks like it was https://github.com/godotengine/godot/issues/92880, but I found a way around it. 

The Dragger Nodes have to be created in the `add_child_notify` so the duplication code sees it, the resort happens too late. This probably also prevents an excess sort children notification happening immediately after a sort.
Note: it is important that the Dragger Nodes are not removed during `remove_child_notify` since that would cause issues when deleting or changing the `SplitContainer`'s type.

Added a compatibility special case for when there are exactly 2 children and they have the `Expand` flag. This is because minimum size now needs to be considered for the positions, but it is not needed when there are only 2 children.

KoBeWi commented on doc/classes/SplitContainer.xml:
```diff
@@ -1,10 +1,10 @@
 	</brief_description>
 	<description>
-		A container that accepts only two child controls, then arranges them horizontally or vertically and creates a divisor between them. The divisor can be dragged around to change the size relation between the child controls.
+		A container that arranges child controls horizontally or vertically and creates divisors between them. The divisors can be dragged around to change the size relations between the child controls.
```
Why did you repeat the above, but with divisors instead of grabbers? 🙃

kitbdev replied:
That's just how it was beforehand. Changed to use grabbers.

KoBeWi commented on doc/classes/SplitContainer.xml:
```diff
@@ -13,19 +13,24 @@
 		<member name="dragger_visibility" type="int" setter="set_dragger_visibility" getter="get_dragger_visibility" enum="SplitContainer.DraggerVisibility" default="0">
 			Determines the dragger's visibility. See [enum DraggerVisibility] for details.
 		</member>
 		<member name="split_offset" type="int" setter="set_split_offset" getter="get_split_offset" default="0">
```
This should be deprecated and the description should mention it's compatibility only.

kitbdev replied:
Deprecated.
Being deprecated already means it's for compatibility, so it doesn't need to be mentioned in the description, right? I don't think other deprecated properties mention it either.

KoBeWi commented on doc/classes/SplitContainer.xml:
```diff
@@ -36,39 +41,39 @@
 	<theme_items>
 		<theme_item name="autohide" data_type="constant" type="int" default="1">
-			Boolean value. If 1 ([code]true[/code]), the grabber will hide automatically when it isn't under the cursor. If 0 ([code]false[/code]), it's always visible.
+			Boolean value. If [code]1[/code] ([code]true[/code]), the grabbers will hide automatically when they aren't under the cursor. If [code]0[/code] ([code]false[/code]), The grabbers are always visible.
```
```suggestion
			Boolean value. If [code]1[/code] ([code]true[/code]), the grabbers will hide automatically when they aren't under the cursor. If [code]0[/code] ([code]false[/code]), the grabbers are always visible.
```

KoBeWi commented on scene/gui/split_container.h:
```diff
@@ -62,13 +63,15 @@ class SplitContainer : public Container {
 private:
-	int split_offset = 0;
-	int middle_sep = 0;
+	Vector<int> split_offsets;
```
```suggestion
	PackedInt32Array split_offsets;
```
I'd replace all usages of `Vector<int>` with this, it's clearer especially for exposed API.

KoBeWi commented on scene/gui/split_container.cpp:
```diff
@@ -153,102 +172,326 @@ Ref<Texture2D> SplitContainer::_get_grabber_icon() const {
+	Point2i position_range = Point2i(0, get_size()[axis]);
+	position_range.x += sep * p_dragger_index;
+	position_range.y -= sep * (dragger_positions.size() - p_dragger_index);
+
+	for (int i = 0; i < (int)dragger_positions.size() + 1; i++) {
```
```suggestion

	int max_dragger_position = (int)dragger_positions.size() + 1;
	for (int i = 0; i < max_dragger_position; i++) {
```

KoBeWi commented on scene/gui/split_container.cpp:
```diff
@@ -153,102 +172,326 @@ Ref<Texture2D> SplitContainer::_get_grabber_icon() const {
+				}
+			}
+		}
+		stretch_complete = true;
```
This is set in every loop iteration, no? So `while` looks redundant.

KoBeWi reviewed (approved):
I did some (not super-extensive) testing and it works correctly. The code looks fine overall, but it's not like I'm very familiar with SplitContainer. Would be nice if @groud can take a look too.

I spotted one *minor* issue that adding a child will add new entry to `split_offsets`, but deleting it won't delete the entry. It doesn't really have adverse effect though.

kitbdev commented:
Fixed the stretching algorithm, it should now match BoxContainer in the case that the not-first-stretching child has a minimum size greater than its desired size.

> I spotted one minor issue that adding a child will add new entry to split_offsets, but deleting it won't delete the entry. It doesn't really have adverse effect though.

Yeah, this is because removing a child and undoing it won't restore the split offset, so I don't resize split_offsets to prevent loss of data. I think the undo system needs more functionality for this, so it can be fixed in the future.

WhalesState commented:
Note:
This will break compatibility with existing projects when loading them, because `set_split_offset()` and `get_split_offset()` have been removed, they should be deprecated instead, setting and getting the first element of the `split_offsets` array if it's size is greater than 0, or returning 0.0 instead.

AThousandShips commented:
They *have* been deprecated, please check the code

WhalesState commented:
Sorry i just noticed the variable was removed, and thought they have missed this, Amazing work and thanks for making this ^^.

WhalesState commented:
I was testing it since i have had a hard time making a similar `Container`, and i have found some issues that we can fix before implementing.

1- Hiding a child makes the other children uses a wrong offset index, so hidden children should be respected when getting the sortable controls.
2- We should not deal with all internal children as draggers, adding an internal control node will not be sorted we better ignore non Control nodes and `SplitContainerDragger` nodes only.
3- Size flags should only be taken into account when calculating offsets for the first time or when the node is inside editor, any other sorting after that should be calculated using `drag_offsets` if the offsets size is greater than the sortable children.

https://github.com/user-attachments/assets/3c8f4b3c-e00c-4df7-9ba9-149e2c19eb0e

kitbdev commented:
Rebased and removed unused includes.

>    Hiding a child makes the other children uses a wrong offset index, so hidden sortable children should be respected when getting the dragger's offset index.

Other Containers ignore hidden children intentionally, so only having enough split offsets for the valid visible children is expected.
This would also cause compatibility issues, since if a user has hidden controls at the start of their SplitContainer it would affect the split offset.

>    We should not deal with all internal children as draggers, adding an internal control node will not be sorted. We better ignore non Control nodes, top level controls and SplitContainerDragger nodes only.

Ignoring internal children was the behavior beforehand, but other Containers work like that so it makes sense to not ignore them all. Since it is not directly related and may need more review this can be done in another PR.

>    Size flags should only be taken into account when calculating offsets for the first time or when the node is inside editor, any other sorting after that should be calculated using drag_offsets if the offsets size is greater than or equal the sortable children size - 1.

This would mean changing size flags after starting does nothing, which doesn't seem very useful. Also what if the user adds a child and then sets a size flag for it after? It won't be recognized. Other containers like BoxContainer also don't do it like this.


Having it saved as a ratio seems nice, but it would break compatibility. 
- There is a related proposal: https://github.com/godotengine/godot-proposals/issues/4033

WhalesState commented:
Let's say we use offsets as a ratio which seems the best solution for this, it won't break compatibility since we can control the old behaviour in the deprecated method by converting old offset `offset / (vertical ? size.y : size.x)`. this will make it easier to deal with offsets and will make it possible to control the offsets in the editor.

Size flags are made for static containers because we can't control their size manually, `SplitContainer` Should be an exception, since it's the only container that allows you to manually control it's children size. We better force all the children to only use the `FILL` flag and we expand them by default when they are created for the first time, ie. we give them all an equal size when children are added or removed and the split offsets are empty or smaller than the draggers size, then we control the offsets manually in the editor or in script.

```cpp
Vector<int> SplitContainer::get_allowed_size_flags_horizontal() const {
	Vector<int> flags;
	flags.append(SIZE_FILL);
	return flags;
}

Vector<int> SplitContainer::get_allowed_size_flags_vertical() const {
	Vector<int> flags;
	flags.append(SIZE_FILL);
	return flags;
}
```

Containers are numbers and characters are the dragger.

```cpp
// [ 1 |a| 2 |b| 3 |c| 4 |d| 5 ]
// [ 2       |b| 3 |c| 4       ]

// Hiding 1 and 5 should make 2 use b offset and 4 expands till the end. dragger_offsets should not be changed at all when hiding/showing children.

// [ 1 |a| 2 |b| 3 |c| 4 |d| 5 ]
// [ 2       |b| 4            ]

// Hiding 3 should make 2 use b offset and 4 expands to the end.

[ 1 |a| 2 |b| 3 |c| 4 |d| 5 ]
[ 2       |b| 4            ]

// To achieve this, we should get the sortable children first and we save them and then we loop them to get the visible children.

// split_container.h
Vector<Control *> children;

// split_container.cpp
void SplitterContainer::sort_children() {
	children.clear();
	Vector<Control *> visible_children;

	for (int i = 0; i < get_child_count(); i++) {
		Control *c = Object::cast_to<Control>(get_child(i));
		if (!c || get_child(i)->is_class("SplitContainerDragger")) {
			continue;
		}
		if (!c->is_top_level_control()) {
			// get all sortable children even if they are hidden.
			children.append(c);
			if (c->is_visible_in_tree()) {
			        // get all visible children that will be sorted.
				visible_children.append(c);
			}
		}
	}
}

// To get the correct offset index.
int index = children.find(c);


```

> > We should not deal with all internal children as draggers, adding an internal control node will not be sorted. We better ignore non Control nodes, top level controls and SplitContainerDragger nodes only.
> 
> Ignoring internal children was the behavior beforehand, but other Containers work like that so it makes sense to not ignore them all. Since it is not directly related and may need more review this can be done in another PR.
> 
> > Size flags should only be taken into account when calculating offsets for the first time or when the node is inside editor, any other sorting after that should be calculated using drag_offsets if the offsets size is greater than or equal the sortable children size - 1.
> 
> This would mean changing size flags after starting does nothing, which doesn't seem very useful. Also what if the user adds a child and then sets a size flag for it after? It won't be recognized. Other containers like BoxContainer also don't do it like this.
> 
> Having it saved as a ratio seems nice, but it would break compatibility.
> 
> * There is a related proposal: [SplitContainer - add percentage option for better scaling and UI responsiveness. godot-proposals#4033](https://github.com/godotengine/godot-proposals/issues/4033)

WhalesState commented:
A workaround for controlling offsets in the editor is by using the `Control.size_flags` to override the offsets in the canvas item editor, check `Node.is_part_of_edited_scene()` since this shouldn't be applied for editor plugins or running scenes.

kitbdev commented:
>Let's say we use offsets as a ratio which seems the best solution for this, it won't break compatibility since we can control the old behaviour in the deprecated method by converting old offset offset / (vertical ? size.y : size.x).

I don't think converting it is enough, since if the size changes, the value changes as well.

> We better force all the children to only use the FILL flag and we expand them by default when they are created for the first time

We can't make SplitContainer not use its children's size flags, that breaks compatibility. Even the editor uses size flags in SplitContainers. It would make the implementation simpler, but it wouldn't have much benefit to the user.
It is also useful, since it is used when resizing the SplitContainer to know which side to expand.

> SplitContainerDragger can be excluded when looping over the children, if (get_child(i)->is_class("SplitContainerDragger")).

Yes, it should be simple to implement, but it is not really related to this PR. It's better to have a separate PR for it.

WhalesState commented:
I have found an issue, by hiding the left or the right dock completly, their draggers are not removed.

https://github.com/user-attachments/assets/ca39c0dc-c764-43f2-aa0c-3cee13d740ec

kitbdev commented:
Draggers are now hidden when there is 0 or 1 child.
They aren't removed since the first one is added in the constructor, and it could cause issues with duplicate.

WhalesState commented:
Great work.
Thanks for making this and good luck with your software ^^.

kitbdev commented:
- Rebased https://github.com/godotengine/godot/pull/90439 on this so it can be tested on the editor ui.

There are some issues when docks are moved (when a child visibility changes), docks can jump around when they get another's split offset. There is also a case where the minimum size isn't respected, so this still needs some work. Maybe invisible children should be considered somehow? or automatically remove / insert a split_offset when children hide/show?

WhalesState commented:
> Maybe invisible children should be considered somehow? or automatically remove / insert a split_offset when children hide/show? We may need to attach some metadata to the children for this, like `TabContainer` does.

In cases like hiding a child, it's dragger offset should be kept but ignored and it's dragger should hide, the next child should use an offset based on it's child index. this will preserve the layout when showing/hiding a child control.

In cases like adding/removing children in game or editor plugin, i see that the layout is reset based on the size flags, which will ignore the current offsets that was set by user, it means that if one child is set to expand, and others are not, all the children that aren't expanding will be shrinked to their minimum size, which will force the user to set the offsets again from start, this seems wrong and offsets set by user should also be respected in game and editor plugins. to fix this [in game and editor plugins only] children should be arranged the first time based on their size flags, then once the offsets are set, they should always be respected and not be overrided by size flags except when expanding, the offsets should be tweaked and not reset.

In cases like resizing, it seems to work as expected, children with size flag `expand` expands equally and others respect their offsets.

Just make some tweaks to the `EditorInterface`, and sort all the docks alongside the center v split container inside one `HSplitContainer` and you will notice those issues that i have mentioned.

kitbdev commented:
- Rebased to fix merge conflicts after https://github.com/godotengine/godot/pull/72680

Still working on the visibility issue,

- Due to the complexity and chance of regressions, I'd like to add some tests first #97371

arkology commented:
@kitbdev any chance targeting 4.5?

kitbdev committed:
Allow SplitContainer to have multiple children

Geometror commented:
What's the state of this PR? I'd like to move this forward :)

kitbdev commented:
Rebased and updated. 
Changing the visibility of children, or adding/removing/moving them, will now update the split offsets so that all children try to remain the same size that they were. This adds more complexity, and going back to a recursive approach might be easier.

- Also updated https://github.com/godotengine/godot/pull/90439 for the editor usage

Misc details
- touch dragger moved to SplitContainerDragger
- valid_children are now kept track of on add/remove/move and visibility change. This way we don't have to loop through all children to find them and the current state is easier to determine.
- When a child is added or removed and there is more than 2 children, the split offset now effectively gets clamped for the sizing calculations.
- I added an index to the `get_split_offset`/`set_split_offset` functions for the cpp, but didn't expose these new ones to gdscript. This is because they are useful in SplitContainerDragger when only one offset needs to change and to avoid needing to change all uses with a single element vector. But it isn't as useful in gdscript as you can just set an index on the `split_offsets` directly.
- A `priority_index` parameter was added to `clamp_split_offset` since there wasn't a way to determine which split should get priority, which is needed when they can overlap.
- Changed the tests to use a small macro and helper functions so that they fail on a line inside of the test case, which is helpful when debugging tests.
- Added more tests for when there are only 2 children, to make sure the split offset behaves correctly in more cases, since it is important for the existing way docks work with SplitContainer.
- Added tests for 3 children, but there are a lot of cases I didn't add tests for.

AThousandShips commented on misc/extension_api_validation/4.5-stable.expected:
```diff
@@ -7,3 +7,9 @@ should instead be used to justify these changes and describe how users should wo
+--------
+Validate extension JSON: JSON file: Field was added in a way that breaks compatibility 'classes/SplitContainer/methods/clamp_split_offset': arguments
+
+Optional argument added for index. Compatibility methods registered.
```
```suggestion
Optional argument added for index. Compatibility method registered.
```

KoBeWi commented on doc/classes/SplitContainer.xml:
```diff
@@ -1,22 +1,23 @@
+			<param index="0" name="priority_index" type="int" default="0" />
 			<description>
-				Clamps the [member split_offset] value to not go outside the currently possible minimal and maximum values.
+				Clamps the [member split_offsets] values to not go outside the possible minimum and maximum values, and to not overlap with each other. Use [param priority_index] to prioritize one split offset when there are overlaps.
```
What does "prioritize" mean here? Is it more likely to get clamped or less, or what?

KoBeWi replied:
It definitely needs to be better clarified, but not sure about the exact wording.

CC @Mickeon

kitbdev replied:
Updated the wording: "Clamps the [member split_offsets] values to ensure they are within valid ranges and do not overlap with each other. When overlaps occur, this method prioritizes one split offset (at index [param priority_index]) by clamping any overlapping split offsets to it."

WhalesState commented on scene/gui/split_container.h:
```diff
@@ -106,24 +122,44 @@ class SplitContainer : public Container {
-	void set_split_offset(int p_offset);
-	int get_split_offset() const;
-	void clamp_split_offset();
+	void set_split_offset(int p_offset, int p_index = 0);
```
```suggestion
	void set_split_offset(int p_index, int p_offset);
```

For consistency with other setters in the engine, the index should come first

kitbdev replied:
There already is another method for compatibility `_set_split_offset_first` where I do that, this one is for use in the editor. The offset is the most important parameter in this case, and the index is 0 in most cases so I wanted to have it as a default parameter.

WhalesState replied:
Okay it doesn't matter that much since it will be rarely used in GDScript.

kitbdev replied:
This function isn't even exposed to GDScript.

Geometror reviewed (approved):
Amazing work!
Tested locally, works flawlessly. Code looks fine as well, there are just two small nitpicks.

(Also needs a quick rebase)

Repiteo commented:
Thanks!
